### PR TITLE
IPFS API Client Implementation

### PR DESCRIPTION
**Pull request by amany9000**
# **Google Summer of Code 2019 Final Report** #

## IPFS API Client ##

For the issue #94.  Added src/Network/Ipfs/Api folder for Client Implementation.  

**(in Network.Ipfs.Api.Api)**  :         
- Created IPFS Api Types, their FromJSON Instances and corresponding getter function for functions under call.
- Created a new Content-Type as IPFS Daemon API doesn't return Charset in Content-Type. 

**(in Network.Ipfs.Api.Ipfs)** :   
- Implemented 84 ipfs api functions.                           
- Added call functions for streams, multipart/form-data and regular types.
- Monadic Abstractions for functions under call.

**(in Network.Ipfs.Api.Stream)** :   
- Created IPFS Api Types, their FromJSON Instances and corresponding getter function for functions under streamCall.

**(in Network.Ipfs.Api.Multipart)** :   
- Created IPFS Api Types - AddObj under multipartCall.

**(in Network.Ipfs.Types.Multihash)** :   
- Added Multihash type.

**(in Network.Ipfs.Api.Test.Key)** :   
- Added unit tests for Ipfs Api.

**Other Tasks Completed:** 
- Haddock Documentation for all functions.
- Analysis of IPFS Commands and API.
- Analysis of Servant.API and Servant.Client Library Code.
- Created Multiple Example Codes on Running HTTP API of IPFS Daemon using Servant and example Codes on Running HTTP API of IPFS Daemon using Network.HTTP.Client.

[**Work Folder**](https://drive.google.com/open?id=1GGFVJ_OwBBHKNd-usET-VRE9y1TnTa7u) Contains :- 

- IPFS API Analysis Document.  
- Example codes and Output Screenshots of using IPFS Daemon API using Servant.
- Example codes and Output Screenshots of using IPFS Daemon API using Network.HTTP.Client.

**IPFS API Function Checklist:** 

**Files**
Regular Files API
- [x] ipfs.add(path, [options], [callback])
- [x] ipfs.cat(ipfsPath, [options], [callback])
- [x] ipfs.get(ipfsPath, [options], [callback])
- [x] ipfs.ls(ipfsPath, [callback])

block
- [x] ipfs.block.get(cid, [options], [callback])
- [x] ipfs.block.put(block, [options], [callback])
- [x] ipfs.block.stat(cid, [callback])

refs
- [x] ipfs.refs(ipfsPath, [options], [callback])
- [x] ipfs.refs.local([callback])

**Graph**
dag
- [x] ipfs.dag.get(cid, [path], [options], [callback])
- [x] ipfs.dag.put(dagNode, [options], [callback])
- [x] ipfs.dag.resolve(cid, [callback])

object
- [x] ipfs.object.data(multihash, [options], [callback])
- [x] ipfs.object.get(multihash, [options], [callback])
- [x] ipfs.object.links(multihash, [options], [callback])
- [x] ipfs.object.new([template], [callback])
- [x] ipfs.object.patch.addLink(multihash, DAGLink, [options], [callback])
- [x] ipfs.object.patch.appendData(multihash, data, [options], [callback])
- [x] ipfs.object.patch.rmLink(multihash, DAGLink, [options], [callback])
- [x] ipfs.object.patch.setData(multihash, data, [options], [callback])
- [x] ipfs.object.diff(multihash, multihash, [callback])
- [x] ipfs.object.put(obj, [options], [callback])
- [x] ipfs.object.stat(multihash, [options], [callback])

pin
- [x] ipfs.pin.add(hash, [options], [callback])
- [ ] ipfs.pin.ls([hash], [options], [callback])
- [x] ipfs.pin.rm(hash, [options], [callback])

MFS (mutable file system) specific 

- [x] ipfs.files.chcid([from, to], [callback])
- [x] ipfs.files.cp([from, to], [callback])
- [x] ipfs.files.flush([path], [callback])
- [x] ipfs.files.ls([path], [options], [callback])
- [x] ipfs.files.mkdir(path, [options], [callback])
- [x] ipfs.files.mv([from, to], [callback])
- [x] ipfs.files.read(path, [options], [callback])
- [x] ipfs.files.rm(path, [options], [callback])
- [x] ipfs.files.stat(path, [options], [callback])
- [x] ipfs.files.write(path, content, [options], [callback])

**Network**
 
bootstrap
- [x] ipfs.bootstrap.add(addr, [options], [callback])
- [x] ipfs.bootstrap.list([callback])
- [x] ipfs.bootstrap.rm(addr, [options], [callback])

bitswap
- [x] ipfs.bitswap.stat([callback])
- [x] ipfs.bitswap.wantlist([peerId])
- [x] Ipfs.bitswap.ledger()
- [x] Ipfs.bitswap.reprovide()

dht
- [x] ipfs.dht.findPeer(peerId, [callback])
- [x] ipfs.dht.findProvs(hash, [callback])
- [x] ipfs.dht.get(key, [callback])
- [x] ipfs.dht.provide(cid, [callback])
- [x] ipfs.dht.query(peer, [callback])

pubsub
- [x] ipfs.pubsub.ls(topic, [callback])
- [x] ipfs.pubsub.peers(topic, [callback])
- [x] ipfs.pubsub.publish(topic, data, [callback])
- [x] ipfs.pubsub.subscribe(topic, handler, [options], [callback])

swarm
- [x] ipfs.swarm.peers([options], [callback])
- [ ] ipfs.swarm.addrs([callback])
- [x] ipfs.swarm.connect(addr, [callback])
- [x] ipfs.swarm.disconnect(addr, [callback])
- [x] ipfs.swarm.filters([callback])
- [x] ipfs.swarm.filters.add(filter, [callback])
- [x] ipfs.swarm.filters.rm(filter, [callback])

name
- [ ] ipfs.name.publish(addr, [options], [callback])
- [ ] ipfs.name.pubsub.cancel(arg, [callback])
- [ ] ipfs.name.pubsub.state([callback])
- [ ] ipfs.name.pubsub.subs([callback])
- [ ] ipfs.name.resolve(addr, [options], [callback])

**Node Management**
 
 miscellaneous operations
- [x] ipfs.dns(domain, [callback])
- [x] ipfs.id([callback])
- [x] ipfs.idPeer(peerId ,[callback])
- [x] ipfs.ping(id, [options], [callback])
- [x] ipfs.shutdown([callback])
- [x] ipfs.version([callback])

cid
- [x] Ipfs.cid.bases()
- [x] Ipfs.cid.codecs()
- [x] Ipfs.cid.hashes()
- [x] Ipfs.cid.base32(cid)
- [x] ipfs.cid.format(cid)

config
- [x] ipfs.config.get([key], [callback])
- [x] ipfs.config.replace(config, [callback])
- [x] ipfs.config.set(key, value, [callback])

stats
- [x] ipfs.stats.bw([options], [callback])
- [x] ipfs.stats.repo([options], [callback])

log
- [x] ipfs.log.level(subsystem, level, [options], [callback])
- [x] ipfs.log.ls([callback])
- [x] ipfs.log.tail([callback])

repo
- [x] ipfs.repo.version([callback])
- [x] ipfs.repo.fsck([options], [callback])
- [x] ipfs.repo.gc([options], [callback])
- [x] ipfs.repo.verify([options], [callback])

key
- [x] ipfs.key.gen(name, [options], [callback])
- [x] ipfs.key.list([options, callback])
- [x] ipfs.key.rename(oldName, newName, [callback])
- [x] ipfs.key.rm(name, [callback])

## WebSocket Layer ##

runWeb3WS is the web3 runner for default WebSocket Provider at 'ws://127.0.0.1:8546'.

**Tasks Completed:**

**(in Network.JsonRpc.TinyClient) :** 
- Extending JsonRpcClient to hold a WebSocket Connection.
- Pattern matching in connection function and implementing the code for WebSocket connection.

**(in Network.Ethereum.Api.Provider) :** 
- Updating runWeb3With function by replacing .= of microlens-mtl by State Monad Functions
- Pattern matching in runWeb3' function and implementing the code for WebSocket connection.
- Adding the getConnection function, which returns a WebSocket connection data type.
- Adding WS.sendClose function to close WebSocket Connection.

[**Work Folder**](https://drive.google.com/open?id=1kYqnOpvd0zDfBd262Em98KcJi55t9AOq) Contains :- 
- WS API Analysis Document.
- Example codes and Output Screenshots of Using websockets library with Geth or Parity.
- Example codes and Output Screenshots of getConnection and runWeb3WS functions respectively.
- The codes also include the Required Geth and Parity Specifications.

**Other Tasks Completed:**
- Created Multiple Example Codes on Running WS API of Geth/Parity using the websockets Library.
- Example code of the getConnection and the runWeb3WS functions respectively.
- Analysis of hs-web3 Library Code.
- Analysis of WebSockets Library Code.
- Analysis of mtl library (focusing on MonadState), MTL Lenses (micorlens-mtl) and Aeson (used for encoding and decoding of JSON RPC Requests and Responses respectively).

### CODE REVIEW

**Comment by akru** [comment]
Hello, @amany9000, thank you for your work!

Could you remake implementation list into check list like that:

- [ ] todo

it's a bit more representative.

And also remark about code, `Text` is preferred of `String` because it's much more memory efficient.

File: src/Network/Ethereum/Api/Provider.hs [hunk 1]
```diff
@@ -51,22 +56,27 @@ instance Exception Web3Error
...
 
 -- | 'Web3' monad runner, using the supplied Manager
 runWeb3With :: MonadIO m
             => Manager
             -> Provider
             -> Web3 a
             -> m (Either Web3Error a)
-runWeb3With manager provider f =
-    runWeb3' provider $ jsonRpcManager .= manager >> f
+runWeb3With manager provider f = do
```
Comment (akru) [inline src/Network/Ethereum/Api/Provider.hs]: @amany9000 Could you please explain this part of code.

File: src/Network/Ethereum/Api/Provider.hs [hunk 2]
```diff
@@ -51,22 +56,27 @@ instance Exception Web3Error
 
 --TODO: Change to `HttpProvider ServerUri | IpcProvider FilePath` to support IPC
 -- | Web3 Provider
-data Provider = HttpProvider String
+data Provider = HttpProvider String | WSProvider String Int
   deriving (Show, Eq, Generic)
 
-instance Default Provider where
```
Comment (akru) [inline src/Network/Ethereum/Api/Provider.hs]: Why default provider couldn't be HttpProvider?

File: src/Network/Ipfs/Api/Api.hs [hunk 3]
```diff
@@ -0,0 +1,802 @@
...
+
+import           Control.Arrow                 (left)
+import           Control.Monad
+import           Data.Aeson
+import           Data.Int
+import           Data.ByteString.Lazy          (toStrict)
+import qualified Data.ByteString.Lazy.Char8()
+import qualified Data.HashMap.Strict           as H
+import           Data.Proxy           
+import qualified Data.Text                     as TextS
```
Comment (akru) [inline src/Network/Ipfs/Api/Api.hs]: So many `TextS` prefixes, why you don't just import `Text` type?

File: src/Network/Ethereum/Api/Provider.hs (hunk 2 above)
Comment (amany9000) [reply src/Network/Ethereum/Api/Provider.hs]: It was done to provide default provider for both HTTP and WebSocket.

File: src/Network/Ipfs/Api/Api.hs (hunk 3 above)
Comment (amany9000) [reply src/Network/Ipfs/Api/Api.hs]: Agreed, I'll change it.

File: src/Network/Ethereum/Api/Provider.hs (hunk 1 above)
Comment (amany9000) [reply src/Network/Ethereum/Api/Provider.hs]: It gave this error - 

   ```
Couldn't match type ‘a0
                           -> Data.Functor.Identity.Identity Manager’
                     with ‘JsonRpcClient’
      Expected type: Lens.Micro.Type.ASetter
                       JsonRpcClient JsonRpcClient a0 Manager
        Actual type: JsonRpcClient -> Manager
```

File: src/Network/Ethereum/Api/Provider.hs (hunk 2 above)
Comment (akru) [reply src/Network/Ethereum/Api/Provider.hs]: Yes, I see, but Default instance could be stayed for HTTP provider.

**Comment by akru** [comment]
@amany9000 Could you please fix https://travis-ci.org/airalab/hs-web3/jobs/585687258

**Pull request merged**